### DISCUSSION TITLE
AST and code generation using current TypeScript version

### DISCUSSION THREAD

**Discussion by mfulton26** [denoland/deno#14906] [Q&A] [answered]
How can I use the TypeScript source parser/compiler that is part of my installed `deno` version to parse a module, look for certain content, and then use that content to generate other source files?

e.g. I want to set up a little tool in my project to watch certain source files and when I make changes to them I want to parse the abstract syntax tree, find any `function` marked with `export default`, if there is one inspect its signature (specifically its name and its parameters but possibly its return type and/or comments too) and use that information to derive and generate other source files to automate some mechanical/mundane work in my project so that I don't have to create a bunch of files by hand that differ only by import paths and some TypeScript types.

I have tools I can reference for parsing (e.g. https://astexplorer.net/) and I'll figure out the transform part next but first I need to get access to the `typescript` module. Is there a way to use the one already part of `deno` or do I need to bring my own and maintain it myself by upgrading it whenever Deno upgrades, etc.

I did some investigating and it looks like [dnt](https://github.com/denoland/dnt) uses https://github.com/dsherret/ts-morph which appears to somehow compile/bundle its own version of the TypeScript compiler ([here](https://github.com/dsherret/ts-morph/blob/latest/deno/common/typescript.js)). I'd prefer to not need to grab `typescript` myself but use what's already available in `deno` but maybe I need to bring my own and figure out how to do that.

**Comment by bartlomieju** [comment, marked as answer]
> How can I use the TypeScript source parser/compiler that is part of my installed deno version to parse a module, look for certain content, and then use that content to generate other source files?

The short answer is no - Deno doesn't expose TSC internals to users for programmatic access.

> I did some investigating and it looks like [dnt](https://github.com/denoland/dnt) uses https://github.com/dsherret/ts-morph which appears to somehow compile/bundle its own version of the TypeScript compiler ([here](https://github.com/dsherret/ts-morph/blob/latest/deno/common/typescript.js)). I'd prefer to not need to grab typescript myself but use what's already available in deno but maybe I need to bring my own and figure out how to do that.

Using `ts-morph` or pulling `typescript` from a CDN is a preferred solution to the problem at hand.

**Reply by mfulton26** [reply]
Thanks!
